src/Simulator/CMakeLists.txt

```diff
@@ -0,0 +1,13 @@
+# Copyright (C) 2020 Istituto Italiano di Tecnologia (IIT). All rights reserved.
+# This software may be modified and distributed under the terms of the
+# GNU Lesser General Public License v2.1 or any later version.
+
+if(BIPEDAL_LOCOMOTION_CONTROLLERS_COMPILE_Simulator)
+
+  add_bipedal_component(
+    NAME                   Simulator
+    SOURCES                src/Simlator.cpp src/Visualizer.cpp
+    PUBLIC_HEADERS         include/BipedalLocomotionControllers/Simulator/Simulator.h include/BipedalLocomotionControllers/Simulator/Simulator.tpp include/BipedalLocomotionControllers/Simulator/Integrator.h include/BipedalLocomotionControllers/Simulator/Integrator.tpp include/BipedalLocomotionControllers/Simulator/Visualizer.h
```

Comment:
> It is better to keep in separate libraries the dynamical system, the integrator, and the visualizer. Separating the components would allow us to reuse some of them while changing the others. For example, we could load only the dynamical system in Matlab and use ode45 to integrate while using some other toolbox for visualization. In the end you can still have another component that reunites them.

Ok for splitting the component into three different components. Probably we will need 4 components (Dynamical system, Integrators, Visualizer, Simulator). The simulator will instantiate the dynamical system and the integrators (right now [we use](https://github.com/dic-iit/bipedal-locomotion-controllers/blob/53b27be78e62fc8e3d5638043f014162fd5b3bee/src/Simulator/include/BipedalLocomotionControllers/Simulator/Simulator.h#L89-L103) 6 integrators)

> As a follow up, be careful of not confusing "simulator" with "dynamical system".

The problem here is that we can control the robot in 3 different ways (torque, acceleration and velocity). What is a system here? Do we need three separate systems?

> The need for changing the integrator is pretty common. I would not restrict it to a single one.

I agree

> I am biased here. iDynTree already has some interfaces for the definition of a dynamical system (https://github.com/robotology/idyntree/blob/master/src/optimalcontrol/include/iDynTree/DynamicalSystem.h) and an integrator (https://github.com/robotology/idyntree/blob/master/src/optimalcontrol/include/iDynTree/Integrator.h). Have you considered using them? If so, what led you to a different path?

- The Integrator was to complex, I was looking for something similar to [this one](http://www.icub.org/software_documentation/classiCub_1_1ctrl_1_1Integrator.html) 
- The same for the dynamical system  

The code was implemented this way in order to provide a prompt response for experiments during the pre-IROS period. However, a future implementation - that unfortunately, I am not planning to provide at the moment - can for sure provide a better separation of the components (UI from computation) and a more clever way to handle different kinds of integrators.

Reply:
> The problem here is that we can control the robot in 3 different ways (torque, acceleration and velocity). What is a system here? Do we need three separate systems?

I don't see how this is related to the separation between "simulator" and "dynamical system". My comment was mainly oriented to the fact that these two concepts are mixed up in the "Simulator" class. A "dynamical system" is whatever gives you the input to the integrator. Roughly, "dynamical system" + "integrator" = "simulator".

> The Integrator was to complex, I was looking for something similar to this one

This sounds like an excuse to me, especially from you :wink: Notice that the only method you would have needed to implement was [``dynamics``](https://github.com/robotology/idyntree/blob/master/src/optimalcontrol/include/iDynTree/DynamicalSystem.h#L81). Every integrator inheriting from https://github.com/robotology/idyntree/blob/master/src/optimalcontrol/include/iDynTree/Integrator.h would have worked from free. Also, if you needed a single step integration, it was just a matter of changing the visibility of [one method](https://github.com/robotology/idyntree/blob/master/src/optimalcontrol/include/iDynTree/Integrators/FixedStepIntegrator.h#L39). 
I just wanted to understand if we could have avoided maintaining yet another version of an integrator.

> The code was implemented this way in order to provide a prompt response for experiments during the pre-IROS period. However, a future implementation - that unfortunately, I am not planning to provide at the moment 

I totally understand the need for writing code in order to have something working quickly. But, then let me know what's your plan here. If you want to merge it quickly, then, it does not even make much sense for me to do an in-depth review.

Reply:
Anyways, consider my comments about using iDynTree classes for the integrator and the model only as suggestions. They are absolutely not blocking for this PR.
Nevertheless, I would strongly suggest isolating as much as possible the model from the integrator. This would ease a lot future developments IMHO.